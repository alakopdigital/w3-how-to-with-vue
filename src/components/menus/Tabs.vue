<template lang="html">
  <div class="tabs" :class="{'is-animated': isAnimated, 'is-vertical': isVertical}">
    <div class="tabs-links">
      <button class="tab-link" :class="{'is-active': currentIndex === index}" v-for="(link, index) of links" @click="toggle($event, link, index)" :key="index" v-text="link"></button>
    </div>
    <div class="tabs-contents">
      <slot></slot>
    </div>
  </div>
</template>

<script>
export default {
  name: 'tabs',
  props: {
    links: {
      type: Array,
      required: true
    },
    isAnimated: Boolean,
    isVertical: Boolean
  },
  data() {
    return {
      currentIndex: 0,
      currentContentEl: null
    }
  },
  mounted() {
    if (this.links && this.links.length !== 0) {
      // get first button element
      const firstBtn = this.$el.querySelector('.tab-link')
      firstBtn.click()
    }
  },
  methods: {
    toggle($event, link, index) {
      if (this.currentContentEl) {
        this.currentContentEl.classList.remove('is-active')
      }

      $event.target.classList.add('is-active')

      const content = this.$el.querySelector(`#${link}`)
      content.classList.add('is-active')

      this.currentContentEl = content
      this.currentIndex = index
    }
  }
}
</script>

<style lang="scss">
.tabs {
  width: 100%;
  overflow: hidden;
  border: 1px solid #ccc;
  display: flex;
  flex-direction: column;

  &.is-vertical {
    flex-direction: row;

    .tabs-links {
      flex-direction: column;
      min-width: 100px;
      flex: 0.3;
      border-right: 1px solid #ccc;
      border-bottom: none;

      .tab-link {
        text-align: left;
      }
    }
    .tabs-contents {
      flex: 1;
    }
  }

  .tabs-links {
    display: flex;
    width: 100%;
    background-color: #f1f1f1;
    border-bottom: 1px solid #ccc;

    .tab-link {
      background-color: inherit;
      border: none;
      outline: none;
      cursor: pointer;
      padding: 14px 16px;
      transition: 0.3s;

      &:hover {
        background-color: #ddd;
      }

      &.is-active {
        background-color: #ccc;
      }
    }
  }

  .tabs-contents {
    .tab-content {
      display: none;
      padding: 6px 12px;
      border-top: none;

      &.is-active {
        display: block;
      }
    }
  }

  &.is-animated {
    .tab-content {
      animation: fadeEffect 1s;
    }
  }
}

@keyframes fadeEffect {
  from { opacity: 0; }
  to { opacity: 1; }
}
</style>
