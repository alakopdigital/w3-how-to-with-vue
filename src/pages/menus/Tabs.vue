<template lang="html">
  <page
    title="Tabs"
    intro="tabs with CSS and JavaScript">
    <sample>
      <tabs
        is-animated
        :links="['London', 'Paris', 'Tokyo']">
        <div id="London" class="tab-content">
          <h3>London</h3>
          <p>London is the capital city of England.</p>
        </div>
        <div id="Paris" class="tab-content">
          <h3>Paris</h3>
          <p>Paris is the capital of France.</p>
        </div>
        <div id="Tokyo" class="tab-content">
          <h3>Tokyo</h3>
          <p>Tokyo is the capital of Japan.</p>
        </div>
      </tabs>
    </sample>
  </page>
</template>

<script>
import Tabs from '@/components/menus/Tabs.vue'

export default {
  name: 'tabs-page',
  components: {
    Tabs
  }
}
</script>

<style lang="scss">
</style>
