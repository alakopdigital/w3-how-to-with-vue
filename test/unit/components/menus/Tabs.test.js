import { mount, renderer } from '#/unit/utils'
import Tabs from '@/components/menus/Tabs.vue'

describe('Tabs.vue', () => {
  it('render default', () => {
    const wrap = mount(Tabs, {
      propsData: {
        links: ['test']
      },
      slots: {
        default: '<div id="test" class="tab-content">test content</div>'
      },
      attachToDocument: true
    })

    expect(wrap.hasClass('tabs')).toBe(true)
    expect(wrap.vm.currentIndex).toEqual(0)

    expect(wrap.find('.tab-link').hasClass('is-active')).toBe(true)

    // tab-content의 is-active는 renderer에서 상태 값이 없기 때문에 볼 수 없다.
    const $test = wrap.find('#test')
    expect($test.hasClass('tab-content')).toBe(true)
    expect($test.hasClass('is-active')).toBe(true)
    expect($test.text()).toEqual('test content')

    renderer.renderToString(wrap.vm, (err, str) => {
      if (err) console.error(err)
      expect(str).toMatchSnapshot()
    })
  })

  it('tab-link@click', () => {
    const wrap = mount(Tabs, {
      propsData: {
        links: ['test1', 'test2']
      },
      slots: {
        default: `
          <div>
            <div id="test1" class="tab-content">test1 content</div>
            <div id="test2" class="tab-content">test2 content</div>
          </div>
        `
      },
      attachToDocument: true
    })

    expect(wrap.hasClass('tabs')).toBe(true)
    expect(wrap.vm.currentIndex).toEqual(0)

    // 첫번째 버튼
    const $btn1 = wrap.find('.tab-link')

    expect($btn1.hasClass('is-active')).toBe(true)

    const $test1 = wrap.find('#test1')
    expect($test1.hasClass('tab-content')).toBe(true)
    expect($test1.hasClass('is-active')).toBe(true)
    expect($test1.text()).toEqual('test1 content')

    const $btn2 = wrap.find('.tab-link:nth-child(2)')
    expect($btn2.hasClass('is-active')).toBe(false)

    // 2번째 탭 클릭
    $btn2.element.click()
    wrap.update()

    expect(wrap.vm.currentIndex).toEqual(1)

    expect($btn2.hasClass('is-active')).toBe(true)

    const $test2 = wrap.find('#test2')
    expect($test2.hasClass('tab-content')).toBe(true)
    expect($test2.hasClass('is-active')).toBe(true)
    expect($test2.text()).toEqual('test2 content')

    // 이전에 선택된 tab의 is-active가 사라짐
    expect(wrap.find('.tab-link').hasClass('is-active')).toBe(false)
    expect(wrap.find('#test1').hasClass('is-active')).toBe(false)

    renderer.renderToString(wrap.vm, (err, str) => {
      if (err) console.error(err)
      expect(str).toMatchSnapshot()
    })
  })
})
