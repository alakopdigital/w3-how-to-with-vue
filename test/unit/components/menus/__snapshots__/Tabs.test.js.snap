// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs.vue props#is-animated 1`] = `
<div class="tabs is-animated">
    <div class="tabs-links"><button class="tab-link is-active">test</button></div>
    <div class="tabs-contents">
        <div id="test" class="tab-content">test content</div>
    </div>
</div>
`;

exports[`Tabs.vue props#is-vertical 1`] = `
<div class="tabs is-vertical">
    <div class="tabs-links"><button class="tab-link is-active">test</button></div>
    <div class="tabs-contents">
        <div id="test" class="tab-content">test content</div>
    </div>
</div>
`;

exports[`Tabs.vue render default 1`] = `
<div class="tabs">
    <div class="tabs-links"><button class="tab-link is-active">test</button></div>
    <div class="tabs-contents">
        <div id="test" class="tab-content">test content</div>
    </div>
</div>
`;

exports[`Tabs.vue tab-link@click 1`] = `
<div class="tabs">
    <div class="tabs-links"><button class="tab-link">test1</button><button class="tab-link is-active">test2</button></div>
    <div class="tabs-contents">
        <div>
            <div id="test1" class="tab-content">test1 content</div>
            <div id="test2" class="tab-content">test2 content</div>
        </div>
    </div>
</div>
`;
